tidb-binlog/driver/reader/kafka_consumer/sarama.go

```diff
@@ -0,0 +1,260 @@
+	}
+	consumer, err := sarama.NewConsumerFromClient(client)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &Sarama{
+		client:    client,
+		consumer:  consumer,
+		topic:     cfg.Topic,
+		partition: cfg.Partition,
+	}, nil
+}
+
+// ConsumeFromOffset implements kafka.Consumer.ConsumerFromOffset
+func (s *Sarama) ConsumeFromOffset(offset int64, consumerChan chan<- *KafkaMsg, done <-chan struct{}) error {
+	log.Info("start consumer on",
+		zap.String("topic", s.topic),
+		zap.Int32("partition", s.partition),
+		zap.Int64("offset", offset))
+	partitionConsumer, err := s.consumer.ConsumePartition(s.topic, s.partition, offset)
+	if err != nil {
+		return errors.Trace(err)
+
+	}
+	defer partitionConsumer.Close()
+
+	for {
+		select {
+		case <-done:
+			log.Info("finish consumer")
+			return nil
+
+		case emsg := <-partitionConsumer.Errors():
+			log.Error("consume partitionConsumer failed:",
+				zap.String("topic", emsg.Topic),
+				zap.Int32("partitiong", emsg.Partition),
+				zap.Error(emsg.Err))
+			continue
+
+		case kmsg, ok := <-partitionConsumer.Messages():
+			if !ok {
+				log.Info("no message from partition consumer")
+				time.Sleep(time.Second)
+				continue
+			}
+			msg := &KafkaMsg{
+				Value:  kmsg.Value,
+				Offset: kmsg.Offset,
+			}
+			consumerChan <- msg
+		}
+	}
+}
+
+// SeekOffsetFromTS implements kafka.Consumer.SeekOffsetFromTS
+func (s *Sarama) SeekOffsetFromTS(ts int64, topic string, partitions []int32) ([]int64, error) {
+	var err error
+	if len(partitions) == 0 {
+		partitions, err = s.consumer.Partitions(topic)
+		if err != nil {
+			log.Error("get partitions from topic failed", zap.String("topic", topic), zap.Error(err))
+			return nil, errors.Trace(err)
+		}
+	}
+
+	return s.seekOffsets(topic, partitions, ts)
+}
+
+// seekOffsets returns all valid offsets in partitions
+func (s *Sarama) seekOffsets(topic string, partitions []int32, pos int64) ([]int64, error) {
+	offsets := make([]int64, len(partitions))
+	for _, partition := range partitions {
+		start, err := s.client.GetOffset(topic, partition, sarama.OffsetOldest)
+		if err != nil {
+			err = errors.Trace(err)
+			return nil, err
+		}
+
+		end, err := s.client.GetOffset(topic, partition, sarama.OffsetNewest)
+		if err != nil {
+			err = errors.Trace(err)
+			return nil, err
+		}
+
+		log.Info("seek offsets in",
+			zap.String("topic", topic),
+			zap.Int32("partition", partition),
+			zap.Int64("start", start),
+			zap.Int64("end", end),
+			zap.Int64("target ts", pos))
+
+		offset, err := s.seekOffset(topic, partition, start, end-1, pos)
+		if err != nil {
+			err = errors.Trace(err)
+			return nil, err
+		}
+
+		log.Info("seek offset success", zap.Int64("offset", offset), zap.Int64("target ts", pos))
+		offsets[partition] = offset
+	}
+
+	return offsets, nil
+}
+
+func (s *Sarama) seekOffset(topic string, partition int32, start int64, end int64, ts int64) (offset int64, err error) {
+	startTS, err := s.getTSAtOffset(topic, partition, start)
```

Comment:
Maybe we can reuse this binary search algorithm for the two types of consumer?

Reply:
this seekOffset can reuse, I'll fix later

Reply:
reuse [here](https://github.com/pingcap/tidb-tools/pull/292/files#diff-ecf9181c20a3d764f44a149a217fcc8cR73) PTAL